### PR TITLE
Fitting of Spectrum1D

### PR DESCRIPTION
**Pull request by brechmos-stsci**
This PR enables the ability to fit a Spectrum1D object based on a list of `astropy.modeling.models` where each model in the list is set with a initial conditions. There is the ability to choose different `astropy.modeling.fitters` as well.

Not mergable. Just a WIP PR.

### CODE REVIEW

File: docs/fitting.rst [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+The output of the fitting routine is another set of models whose parameters contain the result of the fit.
+(There is a corresponding output model for each input model.)
+
+.. figure:: img/fitting_original_fit.jpg
+   :figwidth: 500px
+   :align: center
+
+   Original spectrum (blue) and fitted spectrum (orange).
+
+.. automodpai:: specuilts.fitting
```
Comment (bsipocz) [inline docs/fitting.rst]: typo here

Comment (brechmos-stsci) [reply docs/fitting.rst]: argh.. darn fingers :).  Thanks!  Will fix.

Comment (brechmos-stsci) [reply docs/fitting.rst]: oh, wait, not just 1 but 2 typos in one line!  <coffee> <need coffee>

**Comment by nmearl** [comment]
I'm a bit confused by this approach. Why have the user provide a list of separate models? Why not have the user provide a single compound model? It seems that the `fitmodel`/`_simple` functions assume that the user only wants to _add_ the list of models, but the astropy modeling machinery is super sophisticated, allowing arithmetic and binary operations to compose complicated compound models. It's not clear how this would be supported using this api abstraction. It also seems like the user has no control over the fitting initialization, since the `fitmodel` function only asks for an uninitialized class, and not a fitter instance (and the function doesn't have e.g. a `fitter_kwargs` dict).

Also, the docs seem to give the impression that the returned object is a list of models matching those that were put in, but that's a bit an over simplification, yeah? The result is a model -- a compound model, which the user can query for a list of the composing sub models, or use directly as a simple model representation of the spectrum, or even use as an input into a more complex spectrum model.

From my understanding, this approach feels a little more complicated than simply asking the user to look at the astropy docs for how to compose a compound model, and feed that, along with some spectrum data, into a fitter. I think I might just be missing the bigger picture?

**Comment by nmearl** [comment]
Great! Sorry to jump the gun.

**Comment by brechmos-stsci** [comment]
@eteq In the https://github.com/spacetelescope/dat_pyinthesky/blob/master/pyinthesky_specutils_fitting.ipynb you have the `window` parameter for the fitting. This essentially tells the fitter to do the fit only within that window/region for a model (if multiple models then would have multiple windows, one for each). 

Given the analysis work and discussion of region, I wonder if it is better to rename `window` to `region`.  They are effectively the same thing ("area on which to work").  I think generally they are similar enough that they should be named the same thing for consistency.

**Comment by bsipocz** [comment]
@brechmos-stsci - it seems that this will need a rebase as many unrelated changes has been merged in from master.

**Comment by brechmos-stsci** [comment]
Going down a rabbit hole and need some guidance.  For fitting one could potentially pass in a compound astropy model, but astropy fitting currently can not fit a compound model that has sub-models that has parameters with astropy units. 

So, my plan was to convert the compound model and spectral data to a common set of units, strip the units, do the fitting, add the units back onto the fitted model and then convert fitted model units back to the original units.

So, going down this rabbit hole:

Attempt 1)  I originally was just going take each sub-model of the compound model, create a new sub-model of the same type, then convert and set its parameters.  The problem is for models such as Chebyshev1D one needs to know the 'order' to create a new instance. So, not easy to do model creation in simple, general way given some models have input that define the number of parameters.

Attempt 2) So, rather than *create* a new sub-model I was going to do a `deepcopy()` of each sub-model and convert, strip and set the parameter values. But in doing this method on Gaussian1D, if the input sub-model amplitude has a Quantity parameter then I can't set a non-Quantity value (it errors out).

So, basically, I have gone down this rabbit hole and there has to be a better way to do this. It all comes down to the fact that astropy fitting does not like compound models that have units.

@eteq @stscicrawford @nmearl  @bsipocz  Any ideas on something a little cleaner I could do?

**Comment by nmearl** [comment]
This has been an on going issue, I have a couple of open issues with this. I tackled the same thing when developing spectacle. To avoid this issue, we had to ensure that [every](https://github.com/nmearl/spectacle/blob/master/spectacle/modeling/profiles.py#L168) [model](https://github.com/nmearl/spectacle/blob/master/spectacle/modeling/custom.py#L5) we used provided a `_parameter_units_for_data_units` method. This is because astropy fitters natively try to remove units before fitting, and then will attempt to add them back if it can.

Finally, for some reason with a long winded explanation, I had to create [defined classes](https://github.com/nmearl/spectacle/blob/master/spectacle/core/spectrum.py#L354) for the compound models. This may or may not be necessary, but it seemed to alleviate some issues with astropy dealing directly with fitting a compound model as opposed to just what it thinks is a custom model.

I'll certain comb through your PR and see if I can't offer some extra help if I can.

**Comment by bsipocz** [comment]
compound models are also due to be refactored, but I don't exactly know the timescale and extent of it though, you may want to check it with Nadia and Perry.

**Comment by brechmos-stsci** [comment]
After numerous discussion with Nadia, Tom, etc etc, I now have a working set of code (ie it fits compound models with units).  @bsipocz, understood, but not sure how long that will be to get into astropy.  

This PR is not quite ready, but it passes tests on my machine and I plotted the data and fitted model/compound model and it looks decent:
image

I will need to clean up this PR a bit.

So, as a summary, astropy fitting will not fit a CompoundModel that has units. So, I stripped the units off the compound model and created a new model.  After discussions with several people, it *seems* the only way to get a representation of the submodels *with* operators is to use `model._tree.traverse_postorder()` SO, this uses a private variable (tsk tsk). If someone has a better way I could do this, please tell me. Then, once I have the traversal I can re-create the compound model without units (after conversion to spectral units), fit, and then add the units back on (same type of procedure).  When I add the units back on, the models in the tree of the compound model does not have updated parameters so I have to grab the fitted values from the compound models parameters list. There are many subtleties here.

Tomorrow, I might clean up this entry but I wanted to get all my thoughts down before I go for the day.

**Comment by brechmos-stsci** [comment]
Ok. after long last, many discussions with people about astropy modeling, compound models and the idiosyncrasies I finally have a version of spectra fitting that works (for at least the test cases :) ).  

@stscicrawford @eteq @bsipocz  and anyone else, please take a look through it.  

Here are the caveats:
  * It uses a private variable `._tree` in CompoundModel.  After talking with Perry, Nadia and some others this appears to be the **only** way to get the CompoundModel operators.
     * I looked at removing the units of each parameter of each sub-model in the CompoundModel but if the parameter of a Model is a Quantity it can only be replaced by a Quantity -- so no-go on that one.
      * I also tried creating a new unit-less model and then just placing into the CompoundModel, but CompoundModel does not allow indexed assignment of a new model (e.g., `compound_model[0] = gaussian_unitless_model` does not work)
  * In the future, when astropy fitting has the ability to fit data to CompoundModels, where both have units, then much of the code in fitmodels.py can be removed.
   * Also in the future, when CompoundModels is re-written this will have to be re-addressed.

File: specutils/fitting/continuum.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+from astropy.modeling.fitting import SLSQPLSQFitter
+
+from ..fitting import fit_lines
+from ..manipulation.smoothing import median_smooth
+
+
+__all__ = ['fit_continuum', 'fit_continuum_generic']
+
+
+def fit_continuum_generic(spectrum, model=Chebyshev1D(3), fitter=SLSQPLSQFitter(),
```
Comment (eteq) [inline specutils/fitting/continuum.py]: What about `fit_generic_continuum` instead?  I get the motivation to have parallel names, but in general I think it's better to use "real words" - that is, make the code literally readable as english where practical.

File: specutils/fitting/continuum.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+       * The models in the list of ``model`` are added together and passed as a
+         compound model to the `~astropy.modeling.fitting.Fitter` class instance.
+
+    """
+
+    #
+    # Simple median smooth to remove spikes and peaks
+    #
+
+    spectrum_smoothed = median_smooth(spectrum, 3)
```
Comment (eteq) [inline specutils/fitting/continuum.py]: I suggest having the median filter window be variable?  I.e., just change the 3 to a keyword.

File: specutils/fitting/continuum.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    spectrum : Spectrum1D
+        The spectrum object overwhich the equivalent width will be calculated.
+
+    model: list of `~astropy.modeling.Model`
+        The list of models that contain the initial guess.
+
+    fitmodels_type: str
+        String representation of fit method to use as defined by the dict fitmodels_types.
+
+    window : tuple of wavelengths  (NOT IMPLEMENTED YET)
```
Comment (eteq) [inline specutils/fitting/continuum.py]: maybe add the following:
```
if window is not None or weights is not None:
    raise NotImplementedError('window and weights are not yet implemented')
```
?  After all you never know how long this will stay that way...

File: specutils/manipulation/utils.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+from ..spectra import Spectrum1D, SpectralRegion
+
+__all__ = ['excise_regions']
+
+
+def linear_exciser(spectrum, region):
+    """
+    Basic spectral excise method where the spectral region defined by the
+    2-tuple parameter `region` (start and end wavelengths) will result
+    in the flux between those regions be set to a linear ramp of the
```
Comment (eteq) [inline specutils/manipulation/utils.py]: "regions be set" -> "regions to be set

File: specutils/manipulation/utils.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    # Return a new object with the regions excised.
+    return Spectrum1D(flux=modified_flux*spectrum.flux.unit, spectral_axis=wavelengths,
+                      uncertainty=spectrum.uncertainty,
+                      wcs=spectrum.wcs, unit=spectrum.unit,
+                      spectral_axis_unit=spectrum.spectral_axis_unit,
+                      velocity_convention=spectrum.velocity_convention,
+                      rest_value=spectrum.rest_value)
+
+
+def excise_regions(spectrum, regions, exciser=linear_exciser):
```
Comment (eteq) [inline specutils/manipulation/utils.py]: I don't fully understand the motivation for this function.  Why not just implement *only* `linear_exciser` for now, and have future excisers be their own function?

File: specutils/manipulation/utils.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
+from __future__ import division
+
+import numpy as np
+
+from ..spectra import Spectrum1D, SpectralRegion
+
+__all__ = ['excise_regions']
```
Comment (eteq) [inline specutils/manipulation/utils.py]: probably `linear_exciser` needs to be here since it's part of the public API by way of `excise_regions` accepting it as an argument

File: specutils/fitting/fitmodels.py [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+from astropy.modeling import fitting, Model, models
+
+
+__all__ = ['fit_lines']
+
+
+def fit_lines(spectrum, model, fitter=fitting.SimplexLSQFitter(),
+              exclude_regions=None, weights=None, window=None):
+    """
+    Fit the input models (initial conditions) to the spectrum.  Output will be
```
Comment (eteq) [inline specutils/fitting/fitmodels.py]: What about replacing the first sentence with "Fit the input models to the spectrum. The parameter values of the input models will be used as the initial conditions for the fit."

File: specutils/fitting/fitmodels.py [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+    the same models with the parameters set based on the fitting.
+
+    spectrum, models -> models
```
Comment (eteq) [inline specutils/fitting/fitmodels.py]: Not sure what this line is supposed to be?  Perhaps leftover notes?

File: specutils/fitting/fitmodels.py [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+
+    Parameters
+    ----------
+    spectrum : Spectrum1D
+        The spectrum object overwhich the equivalent width will be calculated.
```
Comment (eteq) [inline specutils/fitting/fitmodels.py]: "overwhich" -> "over which"

File: specutils/fitting/fitmodels.py [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+
+    model: `~astropy.modeling.Model` or list of `~astropy.modeling.Model`
+        The model or list of models that contain the initial guess.
+
+    exclude_regions : list of `~specutils.SpectralRegion`
+        List of regions to exclude in the fitting.
+
+    weights : list  (NOT IMPLEMENTED YET)
```
Comment (eteq) [inline specutils/fitting/fitmodels.py]: perhaps add a `NotImplementedError` if it's not None?

File: specutils/fitting/fitmodels.py [hunk 12]
```diff
@@ -0,0 +1,507 @@
...
+        List of weights to define importance of fitting regions.
+
+    Returns
+    -------
+    models : Compound model of `~astropy.modeling.Model`
+        A compound model of models with fitted parameters.
+
```
Comment (eteq) [inline specutils/fitting/fitmodels.py]: Food for thought: a second return value that I think might be useful:  the *individual* models.  That is, it could be:

```
input_model = continuum + line1 + line2 + line3
fitted_compound_model, (fitcontinuum, fitline1, fitline2, fitline3) = fit_lines(..., input_model...)
```

this is a convenience, of course, as you can just extract the second argument from the compound model anyway.  But is it a one-liner to do that, or not?  If it is maybe we can just put it in the docs as an example... But I'm certain *most* users want the separate components in the end, even if for the actually fitting they want the combined model.

File: specutils/fitting/fitmodels.py [hunk 13]
```diff
@@ -0,0 +1,507 @@
...
+    Notes
+    -----
+       * Could add functionality to set the bounds in
+         ``model`` if they are not set.
+       * The models in the list of ``model`` are added
```
Comment (eteq) [inline specutils/fitting/fitmodels.py]: actually, do we really need to do this, since we can just tell the user they should do ``input_model = continuum + line1 + line2 + line3`` ?  It seems like unnecessary duplication to do that for them.

File: specutils/fitting/fitmodels.py [hunk 14]
```diff
@@ -0,0 +1,507 @@
...
+
+    model: `~astropy.modeling.Model` or list of `~astropy.modeling.Model`
+        The model or list of models that contain the initial guess.
+
+    exclude_regions : list of `~specutils.SpectralRegion`
+        List of regions to exclude in the fitting.
+
+    weights : list  (NOT IMPLEMENTED YET)
+        List of weights to define importance of fitting regions.
+
```
Comment (eteq) [inline specutils/fitting/fitmodels.py]: ``window`` is missing in this docstring...

File: specutils/fitting/fitmodels.py [hunk 15]
```diff
@@ -0,0 +1,507 @@
...
+    #   window = (4*u.Angstrom, 6*u.Angstrom) -> tuple
+    #   window = (4, 6)*u.Angstrom -> Quantity
+    #
+
+    #
+    #  Determine the window if there is one
+    #
+
+    # In this case the window defines the area around the center of each model
+    if window is not None and isinstance(window, (float, int)):
```
Comment (eteq) [inline specutils/fitting/fitmodels.py]: wouldn't it be better to express the window as a `SpectralRegion`?  (either internally or, optionally, by the user?)  That way the `window` specification can be either a SpectralRegion to get an *absolute* position, or a scalar or 2-tuple to mean "relative to the center"

File: specutils/fitting/fitmodels.py [hunk 16]
```diff
@@ -0,0 +1,507 @@
...
+        center = model.mean
```
Comment (eteq) [inline specutils/fitting/fitmodels.py]: Not all models have the same definition of "center".  My instinct is to right a heuristic function to determine the center - it should work for Gaussian, Lorentzian, and Voigt, and also try some obvious ones like "mean" or "center", but other than that just give up and tell the user "sorry, you've used too weird of a model for me to know what you mean".

That said, a valid response is "won't do in this PR, make an issue to track it for later"!

File: specutils/fitting/fitmodels.py [hunk 17]
```diff
@@ -0,0 +1,507 @@
...
+
+        if not isinstance(sub_model, Model):
+            model_out_stack.append(sub_model)
+            continue
+
+        #
+        # Make a new instance of the class.
+        #
+
+        if isinstance(sub_model, models.PolynomialModel):
```
Comment (eteq) [inline specutils/fitting/fitmodels.py]: Alternative solution to this that's a bit more general:  
```
if any([p.default == p.empty for p in inspect.signature(sub_model.__class__).parameters.values() if p.kind==p.POSITIONAL_OR_KEYWORD])
```
You might want to unpack that into multiple lines (😉 ), but it looks to me like it does the job: if a model class has a keyword *without* a default, it will be treated like Polynomial.  I think that's basically the key thing here, right?  That there's no default?

You'd probably then have to programatically pick out the parameter name using the signature to replace the ``sub_model.degree`` in the next line, but that should work.

Comment (eteq) [reply specutils/fitting/fitmodels.py]: (Also falls under "you can assign this to @eteq since it's his crazy idea, though :wink:)

File: specutils/fitting/fitmodels.py [hunk 18]
```diff
@@ -0,0 +1,507 @@
...
+    # there is only one).
+    if compound_model:
+        model_out = _combine_postfix(model_out_stack)
+    else:
+        model_out = model_out_stack[0]
+
+    return model_out, dispersion.value, flux.value
+
+
+def _add_units_to_model(model_in, model_orig, spectrum):
```
Comment (eteq) [inline specutils/fitting/fitmodels.py]: From here on we're getting into real code-sorcery level - nice work!

That said, I don't think I was able to give it a proper review, but I'll trust it since it seems to work... 😨 - hopefully we can clean it up a lot whenever we can get this fixed in Astropy. though.

File: docs/fitting.rst [hunk 19]
```diff
@@ -0,0 +1,90 @@
+===================
+Spectrum Fitting
+===================
+
+Specutils has the ability to fit the flux of a `~specutils.Spectrum1D` object.
```
[the remark this answers is not in the document]
Comment (brechmos-stsci) [reply docs/fitting.rst]: kk

File: docs/fitting.rst [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+`~astropy.modeling.Model` that have initial guesses for each of the parameters. 
+
+The internal functionality uses `~astropy.modeling.fitting` routines.  The flux
+is extracted from the `~specutils.Spectrum1D` and is passed along with a 
+compound model created from the model initial guesses.
+
+Model Fitting
+-------------
+
+The first step is to create a set of models with initial guesses as the parameters. Even
```
[the remark this answers is not in the document]
Comment (brechmos-stsci) [reply docs/fitting.rst]: I added a comment in that a method will be provided in the future but user defined initial guesses are possible now.

File: docs/fitting.rst [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+
+    >>> from astropy.modeling import models
+    >>> from specutils import Spectrum1D
+    >>> from specutils import fitting
+    >>> import astropy.units as u
+    >>> import numpy as np
+    >>> np.random.seed(42)
+
+
+For the purpose of the example, build a ``spectrum1d`` variable that will be used in the fitting:
```
[the remark this answers is not in the document]
Comment (brechmos-stsci) [reply docs/fitting.rst]: k

File: docs/fitting.rst [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+    >>> spectrum1d = Spectrum1D(spectral_axis=wave_um, flux=flux_e1)
+
+
+.. figure:: img/fitting_original.jpg
+   :figwidth: 500px
+   :align: center
+
+   Spectrum to be fit.
+
+Now that there is a ``spectrum1d`` to fit, the real fitting setup must happen.  First create
```
[the remark this answers is not in the document]
Comment (brechmos-stsci) [reply docs/fitting.rst]: k

File: specutils/fitting/continuum.py (hunk 2 above)
Comment (brechmos-stsci) [reply specutils/fitting/continuum.py]: changed

File: specutils/manipulation/utils.py (hunk 6 above)
Comment (brechmos-stsci) [reply specutils/manipulation/utils.py]: It was implemented primarily to get around the multiple region thing.  This way anyone who adds their own excise method (either in here, or passes in their own magical method) does not have to worry about the fact that regions could be a list.  So, all they have to implement is assuming one region.

File: specutils/manipulation/utils.py (hunk 7 above)
Comment (brechmos-stsci) [reply specutils/manipulation/utils.py]: added...

File: specutils/fitting/fitmodels.py (hunk 9 above)
Comment (brechmos-stsci) [reply specutils/fitting/fitmodels.py]: it was.... now deleted.

File: specutils/fitting/fitmodels.py (hunk 8 above)
Comment (brechmos-stsci) [reply specutils/fitting/fitmodels.py]: switched to your version

File: specutils/fitting/fitmodels.py (hunk 10 above)
Comment (brechmos-stsci) [reply specutils/fitting/fitmodels.py]: fixed

File: specutils/fitting/fitmodels.py (hunk 11 above)
Comment (brechmos-stsci) [reply specutils/fitting/fitmodels.py]: added in the raise Error

File: specutils/fitting/fitmodels.py (hunk 12 above)
Comment (brechmos-stsci) [reply specutils/fitting/fitmodels.py]: I think, if anywhere, maybe in the docs. Someone who is using it at this level would probably already know they can just do a `fitcontinuum, fitline1, fitline2, fitline3 = fitted_compound_model`, no?

File: specutils/fitting/fitmodels.py (hunk 13 above)
Comment (brechmos-stsci) [reply specutils/fitting/fitmodels.py]: yes, removed the last comment.

File: specutils/fitting/fitmodels.py (hunk 14 above)
Comment (brechmos-stsci) [reply specutils/fitting/fitmodels.py]: Added, but that brings up another point. `window` should probably be based on SpectralRegion now that we have it.

File: specutils/fitting/fitmodels.py (hunk 15 above)
Comment (brechmos-stsci) [reply specutils/fitting/fitmodels.py]: Yeah, we didn't have SpectralRegion before. I'll go through and make the changes.

File: specutils/fitting/fitmodels.py (hunk 16 above)
Comment (brechmos-stsci) [reply specutils/fitting/fitmodels.py]: 👍

File: specutils/fitting/fitmodels.py (hunk 17 above)
Comment (brechmos-stsci) [reply specutils/fitting/fitmodels.py]: Ok. I would like to suggest this goes into a new PR once this PR is merged.

File: specutils/fitting/fitmodels.py (hunk 18 above)
Comment (brechmos-stsci) [reply specutils/fitting/fitmodels.py]: Sorcery is right.  Phew.

It works for what I give it, maybe another way to review for now is if you can think of other *tests* that should be added to make sure my code works. That way we have a better test base for all this. If you have ideas, feel free to add the ideas here or email/slack and I'll add them in.

**Comment by brechmos-stsci** [comment]
> I'm confused why the exciser is in utils rather than a method on Spectrum1D or SpectralRegion. Or is it that #289 is effectively replacing this anyway and it's just a parallel-PR's issue?

It is the #289 / parallel PR thing... I'll modify to use SpectralRegions

File: docs/fitting.rst [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+
+	>>> # Create the flux array
+    >>> base_flux = g1(wave_um) + g2(wave_um) + g3(wave_um)
+
+	>>> # Create the `specutils.Spectrum1D` object.
+    >>> flux_e1 = base_flux + 100*np.random.random(base_flux.shape)*u.mJy
+    >>> spectrum1d = Spectrum1D(spectral_axis=wave_um, flux=flux_e1)
+
+
+.. figure:: img/fitting_original.jpg
```
[the remark this answers is not in the document]
Comment (brechmos-stsci) [reply docs/fitting.rst]: I'll take a look at this.

File: docs/fitting.rst [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+
+.. code-block:: python
+
+    >>> # And... now we can do the fitting
+    >>> fitted_models = fitting.fit_lines(spectrum1d, fg1+fg2+fg3)
+
+The output of the fitting routine is another set of models whose parameters contain the result of the fit.
+(There is a corresponding output model for each input model.)
+
+.. figure:: img/fitting_original_fit.jpg
```
[the remark this answers is not in the document]
Comment (brechmos-stsci) [reply docs/fitting.rst]: 👍

File: specutils/manipulation/utils.py (hunk 5 above)
Comment (brechmos-stsci) [reply specutils/manipulation/utils.py]: 👍

**Comment by brechmos-stsci** [comment]
> I really like the examples you showed in your comment ...

They are from the tests. I added them into the docs using `..plot`.

**Comment by drdavella** [comment]
Current test failures are unrelated and need to be addressed upstream in https://github.com/astropy/ci-helpers. A workaround in the meantime would be setting `CONDA_VERSION` in `.travis.yml` to something sufficiently recent (like 4.5?).

File: specutils/fitting/fitmodels.py (hunk 12 above)
Comment (eteq) [reply specutils/fitting/fitmodels.py]: Actually... *I* didn't know that.  That's great!  Then yes, I'm happy with that, but *definitely* needs to be in the docs.

File: specutils/manipulation/utils.py (hunk 6 above)
Comment (eteq) [reply specutils/manipulation/utils.py]: Ohh, gotcha.  I wonder if we should note that this is likely to be replaced by the multiple-regions thing?

File: docs/fitting.rst (hunk 19 above)
[the remark this answers is not in the document]
Comment (eteq) [reply docs/fitting.rst]: (done in commit I just pushed up)

File: specutils/manipulation/utils.py (hunk 6 above)
Comment (eteq) [reply specutils/manipulation/utils.py]: After thinking about it again I guess this should be decided in #289 - i.e., that might change it, but better to get this in and deal with it there.

**Comment by eteq** [comment]
I made some doc updates in https://github.com/astropy/specutils/pull/210/commits/3273b6261b00a64752af423ffc7bbc3bd4645eb8 as promised, but other than that (and a series of issues I'll create for enhancements), looks good, and will merge as soon as tests pass.
